haematite_ser/src/wrap.rs

```diff
@@ -0,0 +1,242 @@
+            Self::F32(value) => serializer.serialize_f32(*value),
+            Self::F64(value) => serializer.serialize_f64(*value),
+            Self::Char(value) => serializer.serialize_char(*value),
+            Self::Str(value) => serializer.serialize_str(value),
+            Self::Bytes(value) => serializer.serialize_bytes(value),
+            Self::None => serializer.serialize_none(),
+            Self::Some(value) => serializer.serialize_some(value),
+            Self::Unit => serializer.serialize_unit(),
+            Self::UnitStruct(name) => serializer.serialize_unit_struct(name),
+            Self::UnitVariant(name, variant_index, variant) => {
+                serializer.serialize_unit_variant(name, *variant_index, variant)
+            }
+            Self::NewtypeStruct(name, value) => serializer.serialize_newtype_struct(name, value),
+            Self::NewtypeVariant(name, variant_index, variant, value) => {
+                serializer.serialize_newtype_variant(name, *variant_index, variant, value)
+            }
+            Self::Map(map) => {
+                let map: HashMap<_, _> = map.iter().filter(|(_k, v)| v.allowed).collect();
+                let mut serializer = serializer.serialize_map(Some(map.len()))?;
+                for (key, value) in map.iter() {
+                    serializer.serialize_key(key)?;
+                    serializer.serialize_value(&value.inner)?;
+                }
+                serializer.end()
+            }
+            Self::Seq(values) => {
+                let values: Vec<_> = values.iter().filter(|v| v.allowed).collect();
+                let mut serializer = serializer.serialize_seq(Some(values.len()))?;
+                for value in values {
+                    serializer.serialize_element(&value.inner)?;
+                }
+                serializer.end()
+            }
+            Self::Struct(name, map) => {
+                let map: HashMap<_, _> = map.iter().filter(|(_k, v)| v.allowed).collect();
+                let mut serializer = serializer.serialize_struct(name, map.len())?;
+                for (key, value) in map.iter() {
+                    serializer.serialize_field(key, &value.inner)?;
+                }
+                serializer.end()
+            }
+            Self::StructVariant(name, variant_index, variant, map) => {
+                let map: HashMap<_, _> = map.iter().filter(|(_k, v)| v.allowed).collect();
+                let mut serializer = serializer.serialize_struct_variant(
+                    name,
+                    *variant_index,
+                    variant,
+                    map.len(),
+                )?;
+                for (key, value) in map.iter() {
+                    serializer.serialize_field(key, &value.inner)?;
+                }
+                serializer.end()
+            }
+            Self::Tuple(values) => {
+                let values: Vec<_> = values.iter().filter(|v| v.allowed).collect();
```

Comment:
Doing it like this requires the code that eventually deserializes this to know what permissions it has in order to deserialize it properly. This could be avoided by replacing disallowed values with `None` (`null`) instead.

It is assumed that a tuple's length is publicly known and disclosing it doesn't reveal any private information.

Same for `TupleStruct` and `TupleVariant` below.

Reply:
> [...] in order to deserialize it properly.

how come?